flare_portal/api/tests/test_views.py

```diff
@@ -88,3 +93,89 @@ def test_validation(self) -> None:
+
+
+class ModuleDataAPIView(TestCase):
+    def test_post(self) -> None:
+        experiment: Experiment = ExperimentFactory()
+        module: FearConditioningModule = FearConditioningModuleFactory(
+            experiment=experiment
+        )
+        participant: Participant = ParticipantFactory(experiment=experiment)
+
+        url = reverse("api:fear_conditioning_data")
+
+        json_data = {
+            "participant": participant.participant_id,
+            "module": module.pk,
+            "trial": 1,
+            "rating": 5,
+            "conditional_stimulus": "CSA",
+            "unconditional_stimulus": True,
+            "trial_started_at": parse_datetime("2020-01-01T00:00Z"),
+            "response_recorded_at": parse_datetime("2020-01-01T00:00Z"),
+            "volume_level": 50,
+            "headphones": True,
+        }
+
+        resp = self.client.post(url, json_data, content_type="application/json")
+
+        self.assertEqual(201, resp.status_code)
+
+        # Verify data posted
+        data = module.data.get()
+
+        response_data = {
+            **resp.json(),
+            "trial_started_at": parse_datetime(resp.json()["trial_started_at"]),
+            "response_recorded_at": parse_datetime(resp.json()["response_recorded_at"]),
+        }
+
+        self.assertEqual(response_data, {**json_data, "id": data.pk})
+
+        self.assertEqual(data.participant.participant_id, json_data["participant"])
+        self.assertEqual(data.module_id, json_data["module"])
+        self.assertEqual(data.trial, json_data["trial"])
+        self.assertEqual(data.rating, json_data["rating"])
+        self.assertEqual(data.conditional_stimulus, json_data["conditional_stimulus"])
+        self.assertEqual(
+            data.unconditional_stimulus, json_data["unconditional_stimulus"]
+        )
+        self.assertEqual(data.trial_started_at, json_data["trial_started_at"])
+        self.assertEqual(data.response_recorded_at, json_data["response_recorded_at"])
+        self.assertEqual(data.volume_level, json_data["volume_level"])
+        self.assertEqual(data.headphones, json_data["headphones"])
+
+    def test_validation(self) -> None:
+        # Should not be able to add data for a participant that is not in the
+        # same experiment as the module
+        module: FearConditioningModule = FearConditioningModuleFactory()
+        participant: Participant = ParticipantFactory()
+
+        url = reverse("api:fear_conditioning_data")
+
+        json_data = {
+            "participant": participant.participant_id,
+            "module": module.pk,
+            "trial": 1,
+            "rating": 5,
+            "conditional_stimulus": "CSA",
+            "unconditional_stimulus": True,
+            "trial_started_at": parse_datetime("2020-01-01T00:00Z"),
+            "response_recorded_at": parse_datetime("2020-01-01T00:00Z"),
+            "volume_level": 50,
```

Comment:
We're using floats for volume

Reply:
Changed this to a decimal now. One side effect is that the data is returned via the API as a string e.g. `'0.50'`, but this shouldn't be an issue since the data isn't used by the app.